planetary_computer/sas.py

```diff
@@ -49,22 +49,37 @@ def ttl(self) -> float:
 # Key is the signing URL, value is the SAS token
 TOKEN_CACHE: Dict[str, SASToken] = {}
 
+T = TypeVar("T", str, Asset, Item, ItemCollection)
+
 
-def sign(url: str) -> str:
-    """Sign a URL with a Shared Access (SAS)
+def sign(obj: T) -> T:
+    """
+    Sign all relevant URLs within an object with a Shared Access (SAS)
     Token, which allows for read access.
 
     Parameters
     ----------
-    url (str): The HREF of the asset in the format of a URL.
-        This can be found on STAC Item's Asset 'href' value.
+    obj (T): Any supported object containing one or more URLs to sign. Must be one of:
+             str (a URL), Asset, Item, or ItemCollection
 
     Returns
     -------
-    The signed HREF that permits read access to the asset.
+    The object with all relevant URLs updated to include SAS Tokens
     """
-    link = sign_link(url)
-    return link.href
+    if isinstance(obj, str):
+        link = sign_link(obj)
+        return link.href
+
+    if isinstance(obj, Item):
+        return sign_item(obj)
+
+    if isinstance(obj, Asset):
+        return sign_asset(obj)
+
+    if isinstance(obj, ItemCollection):
+        return sign_item_collection(obj)
+
+    raise TypeError("Invalid type, must be one of: str, Asset, Item, or ItemCollection")
```

Comment:
Thoughts on using [`functools.singledispatch`](https://docs.python.org/3/library/functools.html?highlight=singledispatch#functools.singledispatch) as an alternative to these isinstance checks?

It's 90% a style thing, but in theory 3rd-party code could register an implementation for their own types.

Reply:
Interesting! I wasn't aware of that, but I'll give it some consideration.

Reply:
New to me as well - as things stand any additional sign implementations will be stuck with that `T->T` shape. That might be just fine (perhaps even desirable?). If the freedom to implement `T->?` is desirable, the use of `functools.singledispatch` seems like the ideal solution

Reply:
@TomAugspurger I added a commit that switches over to `singledispatch`. I think in general it's a net positive change, so thanks for the recommendation. The only thing I like better about the previous implementation was the better compile time type checking to ensure the function was only accepting one of the supported types. But as @moradology noted, it restricted `sign` to only work on `T->T` signatures, and we've already encountered a need for a divergence with the ability to sign an `ItemSearch` and get back a signed `ItemCollection`. I think the added flexibility and better ease of use for the user (not needing to call a separate function when dealing with an `ItemSearch`) may be a good enough reason to warrant the loosening of types.

Reply:
Great, thanks.

In principle I like the idea of having `sign` always return the same type as the input. But as you say, I don't know how that would work for an `ItemSearch`.